# Sum_Links files exploration and processing
La idea es levantar atributos para el dataset estandar

* observar estructura de la tabla 
* levantar solo los pares (i,j) donde i & j son users de la telco. Para saber si son users de la teclo es necesario levantar el file del hash_map (user_hash -- > int). Descartando asi el resto de los users.
* extraer features de uso "endemico" para nuestros users de la TelCo

## definiciones utilizadas
* timestamp arranca en 0 segundos para 01/01/2012 00:00am
* La Direction, viene dada relativa al user de la primer columna (incoming si entra y viceversa si es outgoing).
* los calls son un conteo de quantity.
* el time se mide en segundos
* daylight time esta entre [7, 19) horas (exclusivo).
* ignoramos los sms, pues no podemos estar seguros de tener esta info en el dataset Argentino

### el dataset vendria con este header
 {'LineKeyOrigin':hash_object \\\  'LineKeyTarget':np.uint32 \\\  'CallsWeekDayLight':np.uint8 \\\   'CallsWeekNight':np.uint8 \\\  'CallsWeekend':np.uint8 \\\  'TimeWeekDaylight':np.uint8 \\\  'TimeWeekNight':np.uint8 \\\  'TimeWeekend':np.uint8 \\\  'SmsWeekDaylight':np.uint8 \\\  'SmsWeekNight ':np.uint8 \\\  'SmsWeekend':np.uint8} 

In [3]:
import pandas as pd; import numpy as np; import os;import random;
pd.set_option('display.max_rows', 300)
from matplotlib import pyplot as plt
%matplotlib inline
import unicodedata
import os
import random
import time

In [4]:
np.random.seed(2016)

In [10]:
#en user_hash_map va a quedar el mapping de hashes por usuario a su int correspondiente.

#users_hash_map= pd.read_csv('/home/juan/mobility-study/output_user_hash_map.txt.gz',
#                       sep = "|",
#                       header=0,
#                       index_col=0)

In [5]:
#seteamos el lugar de trabajo
rootdir="/grandata/voice/sum_links"
os.chdir(rootdir)
year = "2015"; 
month_start= "04";
#todos estos sum_links terminan 2 meses mas tardes
month_end= "0"+str(int(month_start)+2)
input_file= rootdir +"/sum_links_{y}{ms}_{y}{me}.txt.gz".format(y=year,ms=month_start,me=month_end)

In [12]:
!zcat $input_file | head -n10

LineKeyOrigin|LineKeyTarget|CallsWeekDaylight|CallsWeekNight|CallsWeekend|TimeWeekDaylight|TimeWeekNight|TimeWeekend|SmsWeekDaylight|SmsWeekNight|SmsWeekend
CD5844F7EB2C00C63465E43B44EB9025|3F6FEB3BA73D7E51D906B19F1E0A9099|0|0|0|0|0|0|1|0|0
6AEE0F059A39A7E1CF90B93B5BE8E5AD|3F6FEB3BA73D7E51D906B19F1E0A9099|0|0|0|0|0|0|0|0|3
CD5844F7EB2C00C63465E43B44EB9025|21F587020D72C34131DDD04C0FCD6CFF|0|0|0|0|0|0|0|0|1
CD5844F7EB2C00C63465E43B44EB9025|796D132ECDF757CBCE8FB01DDA3891D6|0|0|0|0|0|0|1|0|0
5E8BADC0F256DEC1ACBB1AB2B77527E9|796D132ECDF757CBCE8FB01DDA3891D6|0|0|3|0|0|203|0|0|0
CD5844F7EB2C00C63465E43B44EB9025|5FA3B22C222AFFF5DFAAE7CFCAEBCA30|0|0|0|0|0|0|0|0|1
6B56A956E1A9E56BE4D90543C8706E76|5FA3B22C222AFFF5DFAAE7CFCAEBCA30|0|0|0|0|0|0|1|0|0
2752F0E0F8226FFF2A3189EB131330AC|46B254E4C1A98CECB4F3D52E66BB836D|1|0|1|196|0|36|0|0|0
2752F0E0F8226FFF2A3189EB131330AC|A954CAD047A827B6DEBA10BD797E7B16|10|4|9|1403|208|1416|0|0|0

gzip: stdout: Broken pipe


## Tests
Veamos cuantos usuarios de un sum_links file dado son users de la telco (por cuestiones de velocidad tomamos solo las filas iniciales)

In [234]:
!zcat "/home/juan/mobility-study/output2.txt.gz" | head -n10

USER,EXPOSED,ANTENNA_ID_0,COUNT_0,ANTENNA_ID_1,COUNT_1,ANTENNA_ID_2,COUNT_2,ANTENNA_ID_3,COUNT_3,ANTENNA_ID_4,COUNT_4,ANTENNA_ID_5,COUNT_5,ANTENNA_ID_6,COUNT_6,ANTENNA_ID_7,COUNT_7,ANTENNA_ID_8,COUNT_8,ANTENNA_ID_9,COUNT_9,EPIDEMIC,MOBILITY_DIAMETER
FFFF78CB080636632B2D1DE7A7BFAB03,False,3415,21,2969,5,1320,3,2471,3,1115,2,-1,65535,-1,65535,-1,65535,-1,65535,-1,65535,False,1063.14053495
FFFF51FDDD55A2E24D074DE30C4798E9,True,1207,15,1171,14,1820,9,583,8,2026,5,1066,4,1831,4,1625,2,401,1,825,1,False,520.065958971
FFFE1EDDA25369CBC467536FA6A787FD,True,1847,22,1753,14,1422,11,1028,10,765,9,3072,9,862,8,2463,7,1416,6,871,5,True,989.597851013
FFFD0A912A959CDFCFFAB2F93BFBA435,True,1407,18,1702,8,1613,6,1704,5,1394,3,1398,3,818,2,958,2,726,1,848,1,False,490.58538164
FFFCB235D6D917DDF350F5B11CCB0EE8,True,17,74,174,33,1123,4,1178,4,173,3,890,2,503,1,1039,1,1228,1,1327,1,True,1613.22560564
FFFC9EC4F662C44D3006BA69C9F0DE85,True,876,19,1737,6,1744,5,326,1,378,1,503,1,1118,1,2216,1,2477,1,2962,1,Tru

In [6]:
user_epidemic = pd.read_csv(
                "/home/juan/mobility-study/output2.txt.gz",
                engine = 'c',
                sep = ',',
                header = 0,
                index_col=0,
                dtype = {'EPIDEMIC':np.bool,
                         'ANTENNA_ID_0':np.int16},
                usecols = ['USER','EPIDEMIC','ANTENNA_ID_0']
                )
## ojo con setear los dtypes para las antennas en np.usigned int pues ahi me transforma todos los -1s en el max del dtype
user_epidemic.index.name = 'LineKeyOrigin'

In [29]:
input_file

'/grandata/voice/sum_links/sum_links_201407_201409.txt.gz'

In [13]:
## leemos de a chunks todo el file y lo agregamos a la tabla con los valores que nos importan

#para filtrar por los users de la telco que estamos mirando
hashes = list(np.unique(user_epidemic.index.values))


#ver el tiempo que tarda
start_time = time.time()



table = pd.DataFrame()
for month_start in ["04","07"]:
    year = "2015"
    print("reading month {m} of {tot_n} months, current time is {t} \n".\
              format(m= month_start,tot_n = 2,t=(time.time()-start_time)))
    #todos estos sum_links terminan 2 meses mas tardes
    month_end= "0"+str(int(month_start)+2)
    input_file= rootdir +"/sum_links_{y}{ms}_{y}{me}.txt.gz".format(y=year,ms=month_start,me=month_end)
    raw_table = pd.read_csv(input_file,
                         sep = "|", 
                         header =0,
                         index_col = None,
                         usecols = ['LineKeyOrigin', 'LineKeyTarget', 'CallsWeekDaylight',
                                    'CallsWeekNight','CallsWeekend','TimeWeekDaylight','TimeWeekNight','TimeWeekend'],
                         chunksize = 5*10**7,    
                        engine = 'c',
                         lineterminator='\n',
                         skipinitialspace=True, 
                        error_bad_lines= False,
                        warn_bad_lines= True,
                        dtype = {'CallsWeekDaylight':np.uint16,'CallsWeekNight':np.uint16,'CallsWeekend':np.uint16,
                                          'TimeWeekDaylight':np.uint16,'TimeWeekNight':np.uint16,'TimeWeekend':np.uint16}
                           )
    numb=0;
    for chunk in raw_table:
        numb+=1
        print("reading chunk #{n} of {tot_n}, current time is {t}".\
              format(n=numb,tot_n = 8,t=(time.time()-start_time)))

        table = table.append(chunk[chunk['LineKeyOrigin'].isin(hashes) & chunk['LineKeyTarget'].isin(hashes)])

#como leo varios meses puede haber dos ocurencias distintas para un par (origin,target)
table = table.groupby(['LineKeyOrigin','LineKeyTarget']).sum()
#vuelvo el indice a como estaba
table.reset_index(inplace = True, level=[0,1])        

print("\n total running time of script is %d " % (time.time() -  start_time))

reading month 04 of 2 months, current time is 0.00210404396057
reading chunk #1 of 8, current time is 138.192387104
reading chunk #2 of 8, current time is 305.699479103
reading chunk #3 of 8, current time is 472.512608051
reading chunk #4 of 8, current time is 641.209093094
reading chunk #5 of 8, current time is 964.172450066
reading chunk #6 of 8, current time is 1084.77642298
reading month 07 of 2 months, current time is 1100.73377705
reading chunk #1 of 8, current time is 1239.01961517
reading chunk #2 of 8, current time is 1408.38448215
reading chunk #3 of 8, current time is 1581.92764211
reading chunk #4 of 8, current time is 1750.17677498
reading chunk #5 of 8, current time is 1923.69535613
reading chunk #6 of 8, current time is 2031.36371708
total running time of script is 2112 


### Errors readinglines
algunas lineas no las pudo procesar el C-Parser porque decia que *Error tokenizing data. C error: Expected11 fields in line 189999971,  saw 21* asi que recorro con el error_bad_lines = $False$ y warn_bad_lines = $True$ para luego chequear a mano con sed i.e.

!zcat \$input_file | sed -n '189999969,\$p' | head

In [9]:
!zcat $input_file | sed -n '189999969,$p' | head

gzip: compressed data not read from a terminal. Use -f to force decompression.
For help, type: gzip -h


### Results mirando si (i,j) $\in$ telco_users x telco_users
* intentamos leer un "sample" de la tabla poinendo usando el parametro skiprows = random_bool_array, pero tardaba muchiimo en seleccionar 10^7 filas
* corrida 1: 3.5M de 10M de pares (i,j) eran telco users
* corrida 2: 0.5M de 10M de pares (i,j) eran telco users
* corrida 3: 1.2M de 20M de pares (i,j) eran telco users
* corrida 4: 1.7M de 20M de pares (i,j) eran telco users
* corrida 5: 0.9M de 20M de pares (i,j) eran telco users

### Sample script for attribute selection

In [14]:
table['LineKeyTarget'].isin(hashes).values.all()and (table['LineKeyOrigin'].isin(hashes)).values.all()

True

## filas que no 'aportan' datos. i.e. todas nulas

In [75]:
relevant_cols = [col for col in table.columns.values \
                 if (col != 'EPIDEMIC' and not('Line' in col) and not('ANTENNA' in col) )]
#imprimimos las filas que tienen todas las columnas ==0 (en los atributos relevantes), i.e. no hay actividad.
#y eliminamos estas filas
null_rows= table[(table[relevant_cols].T ==0).all()].index
table.drop(null_rows,inplace=True)

#### Saving first processing file

In [7]:
output_file = "/home/juan/mobility-study/preoutput_sum_links.txt"

In [13]:
pd.DataFrame.to_csv?

In [82]:
#guardo por las dudas para levantar mas rapido a futuro
table.to_csv(output_file, 
             index = False, 
             header = True,
             float_format='%.0f',
             sep="|")

#### Loading first processing file

In [12]:
!zcat "/home/juan/mobility-study/preoutput_sum_links.txt.gz" | head

LineKeyOrigin|LineKeyTarget|CallsWeekDaylight|CallsWeekNight|CallsWeekend|TimeWeekDaylight|TimeWeekNight|TimeWeekend
00000837B868F926F67BD440A0845C1D|5350FBED91E7DA465D7D436ADD1C48F3|0.0|0.0|2.0|0.0|0.0|45.0
00000837B868F926F67BD440A0845C1D|C88DD6FDDF73AF80517291EEA7339B32|1.0|2.0|1.0|47.0|74.0|64.0
000017368A8964F8C71A9F3A1EAF6160|11509D6701C5280D0D0B2271246424B0|1.0|0.0|0.0|318.0|0.0|0.0
000017368A8964F8C71A9F3A1EAF6160|651D2B4E897DB722F28EDB856F4909FE|2.0|0.0|0.0|165.0|0.0|0.0
000017368A8964F8C71A9F3A1EAF6160|FB36428E265F633EC2227140BDDCCB4D|1.0|0.0|0.0|147.0|0.0|0.0
00001ADC8C5705AB4FD8197B5024C1AA|203B6D6813D904D1E9F39EBB40F00C5D|1.0|0.0|0.0|48.0|0.0|0.0
00001ADC8C5705AB4FD8197B5024C1AA|719FD9FE4C22689FF010BD7C15C28A17|0.0|3.0|0.0|0.0|121.0|0.0
00001ADC8C5705AB4FD8197B5024C1AA|CD4F8B895DBF929F0E3ACA3086469CF6|6.0|2.0|3.0|696.0|86.0|176.0
00001ADC8C5705AB4FD8197B5024C1AA|D828A3318C8C24914FE361221029D3EC|0.0|0.0|1.0|0.0|0.0|42.0

gzip: stdout: Broken pipe


In [19]:
table = pd.read_csv( "/home/juan/mobility-study/preoutput_sum_links.txt.gz",
                              header =0,
                              index_col=None,
                              engine="c",
                              sep="|",
                               #dtype = {'CallsWeekDaylight':np.uint16,'CallsWeekNight':np.uint16,'CallsWeekend':np.uint16,
                                #          'TimeWeekDaylight':np.uint16,'TimeWeekNight':np.uint16,'TimeWeekend':np.uint16}                          
                                )

cols = [col for col in table.columns if not('Line' in col or 'ANTENNA' in col) ]
#levanto todo en float, lo paso a int (excepto los hashes)
table[cols].astype(int,copy=False)

CallsWeekDaylight  CallsWeekNight  CallsWeekend  TimeWeekDaylight  \
0                        0               0             2                 0   
1                        1               2             1                47   
2                        1               0             0               318   
3                        2               0             0               165   
4                        1               0             0               147   
5                        1               0             0                48   
6                        0               3             0                 0   
7                        6               2             3               696   
8                        0               0             1                 0   
9                        0               1             1                 0   
10                       0               1             0                 0   
11                       3               7             5               104   
12                       0               9             5                 0   
13                       0               1             0                 0   
14                       0               2             0                 0   
15                       1               1             0                48   
16                       0               1             4                 0   
17                       2               5             2               181   
18                       1               0             0                80   
19                       0               0             1                 0   
20                       0               0             1                 0   
21                       2               1             0              1015   
22                       2               1             0               134   
23                       3               0             2               235   
24                       0               3             0                 0   
25                       0               0             1                 0   
26                       0               0             2                 0   
27                       0               0             1                 0   
28                       0               0             1                 0   
29                       1               0             0                68   
30                       1               0             0               103   
31                       0               0             1                 0   
32                       0               0             1                 0   
33                       0               1             0                 0   
34                       0               4             0                 0   
35                       0               0             1                 0   
36                       0               1             0                 0   
37                       0               0             1                 0   
38                       0               0             1                 0   
39                       0               1             0                 0   
40                       0               0             1                 0   
41                       0               1             0                 0   
42                       0               0             1                 0   
43                       0               0             2                 0   
44                       1               0             0               519   
45                       0               1             0                 0   
46                       0               0             2                 0   
47                       1               0             0                42   
48                       1               0             0               473   
49                       1               0             0                39   
50                       1     

In [20]:
user_epidemic.columns

Index([u'ANTENNA_ID_0', u'EPIDEMIC'], dtype='object')

In [21]:
#agrego la info de epidemicidad de los targets.
table = table.join(user_epidemic[['EPIDEMIC','ANTENNA_ID_0']], on='LineKeyOrigin',how='left',rsuffix='_Origin')
table = table.join(user_epidemic['ANTENNA_ID_0'], on='LineKeyTarget',how='left',rsuffix='_Target')


True

In [171]:
start_time =time.time()
test =  table.sample(10**5)
test.sort_values(by='LineKeyOrigin',inplace=True)
(~test['EPIDEMIC'].isnull().values).all()
time.time() - start_time

1.571415901184082

## Agrego en un .sum() la info de sum_links por antenna y por KeyOrigin

In [97]:
test.head()

LineKeyOrigin                     LineKeyTarget  \
42  000075A3DC589DAD51DB5E3D4EA096DD  3F53782075B7D38D02F35148A5C309C0   
43  000075A3DC589DAD51DB5E3D4EA096DD  446DA7D022DDE45692C01240FBBC3FED   
40  000075A3DC589DAD51DB5E3D4EA096DD  2B55CEE62F71B48CD1F390BD22AA6233   
89  0000999AA2E8173C1E10376BF6C11FF9  19C0F005AA5BA6FABB88F443098D9E94   
91  0000A2185C3D93E2D52D607ADAAFF202  068F2B8B7A6DE1A7CB4EE28A47776652   

    CallsWeekDaylight  CallsWeekNight  CallsWeekend  TimeWeekDaylight  \
42                  0               1             0                 0   
43                  0               0             1                 0   
40                  0               0             1                 0   
89                  1               0             0                51   
91                  1               0             0                45   

    TimeWeekNight  TimeWeekend EPIDEMIC  ANTENNA_ID_0  ANTENNA_ID_0_Target  
42            426            0    False          3152                 1470  
43              0          592    False          3152                 2649  
40              0          147    False          3152                  236  
89              0            0     True          1059                 1844  
91              0            0     True          1805                 1805

In [23]:
cols = [col for col in table.columns if not('Line' in col or 'ANTENNA' in col) ]
user_cols = [col for col in table.columns if ('Line' in col ) ]
antenna_table = table.groupby(['ANTENNA_ID_0','ANTENNA_ID_0_Target'],sort=False)[cols].sum()
antenna_table.drop('EPIDEMIC',inplace=True,axis=1)
#test  =  test.groupby(['LineKeyOrigin','EPIDEMIC'],sort=False)[cols].sum()
antenna_table.reset_index([0,1],inplace=True)
antenna_table.sort_values(by='ANTENNA_ID_0',inplace=True,ascending=True)

In [24]:
# reagrego la informacion de las antennas epidemicas
antennas= pd.read_csv('/home/juan/mobility-study/antennas_mexico.csv',sep = "|",header=0,index_col=None)
antenna_table = antenna_table.join(antennas['EPIDEMIC'], on='ANTENNA_ID_0_Target',how='left')

### Make Antenna_users
Queremos ver cuantos users interactuando con una antenna son Vulnerables y cuantos Totales
Si pensamos que las antennas definen una particion de los nodos del grafo:

Aca Total_users no significa cuantos usuarios residen en la antenna, ni cuantos usuarios de afuera llamaron a la antenna, sino que especificamente habla del # de aristas del grafo social (donde arista implica algun tipo de comunicacion y no considera el tipo de comunicacion) que entran a la antenna i.e. a esta particion 

Idem para Vuln_users, pero ahora cuenta las aristas que provienen exclusivamente de usuarios "epidemicos".

In [25]:
antenna_users = table.groupby(['ANTENNA_ID_0','ANTENNA_ID_0_Target'],sort=False)[user_cols].count()
antenna_users.reset_index(level = [0,1],inplace=True)
antenna_users.sort_values(by='ANTENNA_ID_0',inplace=True,ascending=True)
antenna_users = antenna_users.join(antennas['EPIDEMIC'], on='ANTENNA_ID_0_Target',how='left')
antenna_users = antenna_users.groupby(['ANTENNA_ID_0','EPIDEMIC'])[user_cols].count()
antenna_users.reset_index(level =1 ,inplace=True)
antenna_users.drop('LineKeyTarget',inplace=True,axis=1)
##OJO QUE existen antennas que no tienen contacto con epidemic ==0 o epidemic ==1 entonces aparecen NaNs que 
#deberian ser ==0
antenna_users = antenna_users.pivot(columns='EPIDEMIC')
antenna_users.fillna(value=0,inplace=True)
antenna_users.columns = ['TOTAL_USERS','VULN_USERS']
antenna_users.index.name='ANTENNA_ID_0'
antenna_users['TOTAL_USERS'] = antenna_users['TOTAL_USERS'] +  antenna_users['VULN_USERS']

### Sanity checking im not losing users
hicimos una comparacion de la tabl antes y despues de la ultima linea de codigo y da que
len(np.unique(table.LineKeyOrigin.values)) == 640647
len(np.unique(test.index.levels[0].values)) == 640647

In [27]:
#con esta funcion creamos las filas que vienen a agrupar las sumas de llamados para c/ usuario a partir de contactos con 
#usuarios epidemicos y la suma sin segregar(i.e. sumando los valores para epi==0 y epi ==1)

def get_epidemic_agregator(group):
    #genero la lista vacia de output sin epidemic
    
    labels = [col for col in group.columns.values if (col != 'EPIDEMIC' and not('Line' in col) and not('ANTENNA' in col) )]
    all_labels = labels + [col+"_EPI" for col in labels]
    medio = len(labels)
    vals = [0]*2*(medio) 
    vals = np.array(vals,dtype=np.uint16)
    #total_user = 
    #users = 
    #aca hay que tener mucho cuidados con los casos epidemic solo ==0, solo epi==1, y solo epi ==1 and 0
    if group.shape[0]>1:
        users = group[group['EPIDEMIC']==1]
        vuln_users = group[group['EPIDEMIC']==1]
        grouped = group.groupby('EPIDEMIC')[labels].sum()
        #en el caso que el grupo tenga dos estados de EPIDEMIC, sigue siendo un dataframe
        if grouped.shape[0]>1:
            non_epi = grouped.iloc[0].values
            epi = grouped.iloc[1].values
    #en el caso que el grupo tenga solo 1 estado de EPIDEMIC,colapasa a una serie con la info en el indice
        else:
            if grouped.index.values==0:
                non_epi = grouped.values[0]
                epi = np.zeros(medio)
            if grouped.index.values==1:
                epi = grouped.values[0]
                non_epi = np.zeros(medio)
        
        
        #epi = np.array(group.iloc[1,:].values.tolist()[2:],dtype=np.uint16) if group.iloc[1,:].EPIDEMIC ==1 else \
        #                np.array(group.iloc[0,:].values.tolist()[2:],dtype=np.uint16)
        #non_epi = np.array(group.iloc[0,:].values.tolist()[2:],dtype=np.uint16) if group.iloc[0,:].EPIDEMIC ==0 else \
        #                    np.array(group.iloc[1,:].values.tolist()[2:],dtype=np.uint16)
        
        
        epi =  np.array(epi.tolist(),dtype=np.uint16)
        non_epi =  np.array(non_epi.tolist(),dtype=np.uint16)
        #print("epi es %s" %epi)
        #print("non_epi es %s" %non_epi )
        vals[:medio] += epi.astype(np.uint16) + non_epi.astype(np.uint16)
        vals[medio:] = epi
        #print(vals)

    #caso solo una fila
    elif group.shape[0]==1:

        if group.EPIDEMIC.values==0:
            #primer mitad es valores de la fila, segunda mitad 0
            vals[:medio] += np.array(group[labels].iloc[0].values.tolist(),dtype=np.uint16)

        if group.EPIDEMIC.values==1:
          
            #primer mitad es 0 , segunda mitad valores de la fila
            vals[medio:] += np.array(group[labels].iloc[0].values.tolist(),dtype=np.uint16)
    
    #print("vals object is %s" %vals)
    
    epi_table = pd.DataFrame(data = [vals], columns=all_labels) 
    
    return epi_table


## get_epidemic_aggregator apply on antennas Table

### Testing get_epidemic_aggregator
Corrimos este algoritmo con un sample de mil lineas y comparamos especificamente para aquellos usuarios que tenian caracteristicas de EPIDEMIC==1, con lo cual testeamos a ver si estaba resultando bien o no el siguiente apply en estos casos cruciales (los otros eran mas simples).

In [136]:
antenna_table.head()

ANTENNA_ID_0  ANTENNA_ID_0_Target  CallsWeekDaylight  CallsWeekNight  \
1232115             1                  973                 20               0   
395937              1                 3858                  0               0   
395936              1                 2641                  0               1   
318028              1                   77                  0               1   
853768              1                 2854                  0               0   

         CallsWeekend  TimeWeekDaylight  TimeWeekNight  TimeWeekend  EPIDEMIC  
1232115             0              2733              0            0         0  
395937              1                 0              0          476         0  
395936              0                 0            202            0         1  
318028              0                 0             74            0         1  
853768              1                 0              0           47         1

In [27]:
i = random.randint(0,9200)
antenna_table.iloc[i:i+4,].groupby('ANTENNA_ID_0').apply(get_epidemic_agregator)

CallsWeekDaylight  CallsWeekNight  CallsWeekend  \
ANTENNA_ID_0                                                      
11           0                 49              28            18   

                TimeWeekDaylight  TimeWeekNight  TimeWeekend  \
ANTENNA_ID_0                                                   
11           0              4465           3990         1183   

                CallsWeekDaylight_EPI  CallsWeekNight_EPI  CallsWeekend_EPI  \
ANTENNA_ID_0                                                                  
11           0                     34                  10                 4   

                TimeWeekDaylight_EPI  TimeWeekNight_EPI  TimeWeekend_EPI  
ANTENNA_ID_0                                                              
11           0                  1641                361              170

In [28]:
antenna_table.iloc[i:i+4,]

ANTENNA_ID_0  ANTENNA_ID_0_Target  CallsWeekDaylight  CallsWeekNight  \
36978            11                 2985                  6               3   
18947            11                 1955                  0               0   
30970            11                 1453                 15              18   
31988            11                  230                 28               7   

       CallsWeekend  TimeWeekDaylight  TimeWeekNight  TimeWeekend  EPIDEMIC  
36978             0               300            145            0         1  
18947             1                 0              0           51         0  
30970            13              2824           3629          962         0  
31988             4              1341            216          170         1

In [28]:
start_time = time.time()
antenna_table=antenna_table.groupby('ANTENNA_ID_0').apply(get_epidemic_agregator)
antenna_table.reset_index(level=1,inplace=True)
antenna_table.drop('level_1',axis=1,inplace=True)
time.time()-start_time

24.560290813446045

In [45]:
verify= antenna_table.sample(20)
ind = verify.index.values
ind

array([1354, 3710,  942, 3121, 3686, 1316, 3715,  290, 1914,  175,  136,
       2191,  602,  501, 3020, 2105, 2235,  686, 1358, 1449])

In [279]:
user = 1354
antenna_table[antenna_table['ANTENNA_ID_0'] == user]

ANTENNA_ID_0  ANTENNA_ID_0_Target  CallsWeekDaylight  CallsWeekNight  \
1567          3402                 1378                  1               0   
7990          3402                 3227                  1               0   
7596          3402                 3402                  0               0   

      CallsWeekend  TimeWeekDaylight  TimeWeekNight  TimeWeekend  EPIDEMIC  
1567             0                32              0            0         0  
7990             0                24              0            0         0  
7596             2                 0              0           36         0

In [280]:
verify.loc[user]

CallsWeekDaylight         2
CallsWeekNight            0
CallsWeekend              2
TimeWeekDaylight         56
TimeWeekNight             0
TimeWeekend              36
CallsWeekDaylight_EPI     0
CallsWeekNight_EPI        0
CallsWeekend_EPI          0
TimeWeekDaylight_EPI      0
TimeWeekNight_EPI         0
TimeWeekend_EPI           0
Name: 3402, dtype: int64

In [29]:
antenna_table.reset_index(level=0,inplace=True)
antenna_table = antenna_table.join(antennas['EPIDEMIC'], on='ANTENNA_ID_0',how='left')
antenna_table.head()

ANTENNA_ID_0  CallsWeekDaylight  CallsWeekNight  CallsWeekend  \
0             1               6539            3189          3169   
1             2               6336            3598          3420   
2             3               8958            4688          4118   
3             4               4360            2782          2719   
4             5               2807            1610          1787   

   TimeWeekDaylight  TimeWeekNight  TimeWeekend  CallsWeekDaylight_EPI  \
0             53972          42050        44656                   3989   
1             19025          47471        16603                   3559   
2             43640           5215        38762                   4545   
3             49870          65048        54122                   2105   
4             38676          47700        11503                   1420   

   CallsWeekNight_EPI  CallsWeekend_EPI  TimeWeekDaylight_EPI  \
0                1832              1760                 14042   
1                1852              1893                 11533   
2                2234              2187                 64471   
3                1563              1415                 16605   
4                 899               906                   336   

   TimeWeekNight_EPI  TimeWeekend_EPI  EPIDEMIC  
0              22740            18573         1  
1              50915             3744         1  
2              41947            48825         1  
3               7948            39523         1  
4              13666            15146         1

In [31]:
#finalmente pego la info de antenna_users a la tabla
antenna_table = antenna_table.join(antenna_users[['TOTAL_USERS','VULN_USERS']], on='ANTENNA_ID_0',how='left')

In [35]:
antenna_table[antenna_table['EPIDEMIC']==0].shape

(2794, 16)

In [32]:
antennas.shape, antenna_table.shape

((4429, 5), (4171, 16))

In [33]:
#guardo la info de antennas
antenna_table.to_csv('/home/juan/mobility-study/mexico_antennas_vulns.txt', 
            index = False, 
             header = True,
             sep="|")

In [252]:
antenna_table.shape

(4171, 14)

## get_epidemic_aggregator apply on users Table

### Para optimizar el apply:
mejor correrlo directamente sobre usuarios que tengan algun epidemic ==0 pues estos van a dar resultados diferentes mientras que los otros usarios simplemente "extenderian" la tabla con 0s hacia las columnas _EPIDEMIC . 

En algunos casos el 80% de la tabla son usuarios que NO tienen info epidemica.

In [76]:
def transformation(x,new_cols):
    if x.LineKeyOrigin.values[0] in only_non_epidemic_users:
        for col in new_cols:
            x[col]=0
        x.drop('EPIDEMIC',inplace=True,axis=1)
        #paso el LineKeyOrigin a indice y dropeo esa columna
        x.index = subset['LineKeyOrigin'].values
        x.index.name = 'LineKeyOrigin'
        x.drop('LineKeyOrigin',inplace=True,axis=1)

    return x

In [78]:
new_cols =[col+"_EPI" for col in test.columns.values \
           if (col != 'EPIDEMIC' and not('Line' in col) and not('ANTENNA' in col) )]

grouped.transform(lambda x: transformation(x,new_cols))

/home/juan/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
/home/juan/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
/home/juan/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


LineKeyOrigin                     LineKeyTarget  \
3        000017368A8964F8C71A9F3A1EAF6160  651D2B4E897DB722F28EDB856F4909FE   
29                                    NaN  00A1B253884AFB0C50D2FBF4043E74BC   
71                                    NaN  C89EDEF4E4CEDBB40FF39FE81EBD8687   
69                                    NaN  C0E74CEE3A1693248BDC3FBE3A200D90   
103                                   NaN  68AEB2016C3E272BCC292CD007BA8566   
217                                   NaN  1F79850EAB29B4FC60BABE17D66E3C1C   
248                                   NaN  BDC513825E4591A891E7B8B6A66E4913   
297                                   NaN  D6758363765882542E2C9E768C4710DE   
335      0003ABFB87DA27F8998F8727037B2C7F  AB1AC07646365B3CC52ECF34CC5B7256   
344                                   NaN  587B204530BD3EFF57D44A8CF5A20B0D   
367                                   NaN  6BF6B97544EF04B9393BE1FE9D9969D2   
444                                   NaN  84A624B4BF29EB4488D304DD342CCB59   
477                                   NaN  98CDBBF9168DED054A19BF92EF23FA0F   
554                                   NaN  3C876D371668714C48B6B4CEB437FF27   
636                                   NaN  23DCAFA03FAAEFCBCC0691B1D31C0137   
643      00082D0451B04403DA8214C00B992149  E13A979CBFBFFF70E1B9C3B8A18FB840   
724      000964B68F93458158E652C896880892  0B18E9FB2360088EC202A0438A82B359   
735                                   NaN  E59E3981395543F9EB701E4E7F49FF42   
776      000A6918E549FBAC0C6873CF9446740D  FDCE26A16CB70FFAF6E97BFE15D5FF4F   
801                                   NaN  BC387D4BAC95155967AFFF8DF5782746   
813                                   NaN  0309281DE4BB11B7BEB2F7EDD250A930   
838                                   NaN  92041E46DB49D885C84C317F0E82F18A   
954      000D230C27E0385C8F7F10FBCA62568A  B865C5C8FAA5BB538B73D744FBE4C764   
1046                                  NaN  940690F5385798242F5F7F8107C1DCC0   
1159                                  NaN  68D44932154E0B153DCED749BCA84BD5   
1244                                  NaN  4C29A284EF6277C1D8791D98C11D9670   
1309                                  NaN  F285E62FDE5CF2F53457CB32E1E3C93A   
1327                                  NaN  8501832BCF297F694B72F83F148B0A6F   
1353                                  NaN  BC13A245F30F93E798B1D70A45A70000   
1370     00133E833A767C09631C152A6173DB59  BD85B57A7299ADAE5D24207E956C2B6A   
1430     00136079F64B147B21A745A969BEDE7F  2FF7182631EFEB142504E272BAF70321   
1490     0013DB93302BD78619913D15A046A789  DD612EC28ABADB967E51DB1785B64AAC   
1500     0013F65631325177C376150985798A7A  BCFF1C8DBD372F906AB940C2A69A0EDC   
1544                                  NaN  37D04FB2810C5CC62C7227AEAB70B400   
1587     0015080A2FE1F114DA69F80A56C869A2  AD29A88FF90608BD22CF7823C7CD930E   
1585     0015080A2FE1F114DA69F80A56C869A2  A07F86F969BD0C632F5921439B8962FA   
1594     00152B6A1EB024BB8B2E78431A016734  0B564027D092096F4D0AA19E6379400A   
1617     0015662CA24AD3B7BF1A866DB9C6B439  EA1CD55162AA1B04851C9BB56E6C3DE1   
1712     00164B0553580EE1F24955767D5FB17E  38081581E929543BB3E7E53A3436A6CD   
1804     0016D8D487CE392614E7FCC773B2B7CC  820379D1138182BB0E54E1E05996A1A0   
1832     00174F8F5F2D827DB8D10089E966F494  147B8D84D4147E42D3B8A49D8EC4A02C   
1927                                  NaN  D02D056177C0D91CCAEDE72AB27B62B7   
1971     001953BF301F04F6B1B2163B5A17C6E5  1E1AC28E0E7107468E68D70AC498899D   
2086     001A7F008E6AD1D08742E3820BA12BC7  D6461D1C4D6B4DA01AD33E656EE14F54   
2112                                  NaN  D4FAAA36DB364D39CAE8FC01B3335ECD   
2132                                  NaN  9C332777E94A2A34374C9104DB45B1C8   
2401                                  NaN  B9A8ABF833257FC04906EED39A5396A3   
2435                                  NaN  8D0C069D9F7313FD42E0E5464F40FE0C   
2454                                  NaN  C937E02C1ED45413ACA34E15A68E3D82   
2481                                  NaN  31B8A60486D4C314513FCB7113D783BA   

In [60]:
for name, subset  in grouped:
    #print(name == (True))
    print(name)
    print(type(subset))
    #if name == str(True):

False
<class 'pandas.core.frame.DataFrame'>
True
<class 'pandas.core.frame.DataFrame'>


In [45]:
new_cols =[col+"_EPI" for col in test.columns.values \
           if (col != 'EPIDEMIC' and not('Line' in col) and not('ANTENNA' in col) )]

for name, subset  in grouped:
    print(name)
    if name == True:
        for col in new_cols:
            subset[col]=0
            subset.drop('EPIDEMIC',inplace=True,axis=1)
            #paso el linekeyorigin a indice y dropeo esa columna
            subset.index = subset['LineKeyOrigin'].values
            subset.index.name = 'LineKeyOrigin'
            subset.drop('LineKeyOrigin',inplace=True,axis=1)

False


True




In [ ]:
def is_pure_nonepidemic_user(row,list_of_nonepidemic):
        return row['LineKeyOrigin'] in list_of_nonepidemic

## ojo el proximo llamado en memoria
Es medio peligroso porque copy hace una copia REAL de los datos y no una REFERENCIA

Con esto nos sacamos de encima el problema posterior del *SettingWithCopyWarning*

In [ ]:
start_time =time.time()
test =  table.sample(10**5)
test.sort_values(by='LineKeyOrigin',inplace=True)
(~test['EPIDEMIC'].isnull().values).all()
time.time() - start_time

In [172]:
#al conjunto de los que tienen epi==0 le saco los que tienen epi ==1 me quedo solo con usuarios "puros"
only_non_epidemic_users = list(set(test[test['EPIDEMIC']==0].LineKeyOrigin.values) - set(test[test['EPIDEMIC']==1].LineKeyOrigin.values))
#test 2 se lleva los usuarios mixtos y test3 los "puros" non_epi
test2 = test[~test['LineKeyOrigin'].isin(only_non_epidemic_users)].copy()
test3 = test[test['LineKeyOrigin'].isin(only_non_epidemic_users)].copy()

(3153494, 11)

#### esta corrida abajo tarda 6 mins para 180k users
4 mins para 40k users

80s para 40k users

In [ ]:
whos

In [198]:
test.head()

ANTENNA_ID_0  ANTENNA_ID_0_Target  \
LineKeyOrigin                                                         
00000837B868F926F67BD440A0845C1D          1300                 1300   
00001ADC8C5705AB4FD8197B5024C1AA          1360                 3051   
00001C544BEBBB9109CE3552B4CDC490          3454                 3079   
00002AFB398A29AE5AA1729B78DE22EB          6996                 3715   
000049FDC90665F2A02CCD1F5FABD4D8          1056                 2769   

                                  CallsWeekDaylight  CallsWeekDaylight_EPI  \
LineKeyOrigin                                                                
00000837B868F926F67BD440A0845C1D                  1                      0   
00001ADC8C5705AB4FD8197B5024C1AA                  7                      0   
00001C544BEBBB9109CE3552B4CDC490                  0                      0   
00002AFB398A29AE5AA1729B78DE22EB                  3                      0   
000049FDC90665F2A02CCD1F5FABD4D8                  0                      0   

                                  CallsWeekNight  CallsWeekNight_EPI  \
LineKeyOrigin                                                          
00000837B868F926F67BD440A0845C1D               2                   0   
00001ADC8C5705AB4FD8197B5024C1AA               5                   0   
00001C544BEBBB9109CE3552B4CDC490               1                   0   
00002AFB398A29AE5AA1729B78DE22EB              16                   0   
000049FDC90665F2A02CCD1F5FABD4D8               0                   0   

                                  CallsWeekend  CallsWeekend_EPI  \
LineKeyOrigin                                                      
00000837B868F926F67BD440A0845C1D             3                 0   
00001ADC8C5705AB4FD8197B5024C1AA             4                 0   
00001C544BEBBB9109CE3552B4CDC490             0                 0   
00002AFB398A29AE5AA1729B78DE22EB            10                 0   
000049FDC90665F2A02CCD1F5FABD4D8             1                 0   

                                  TimeWeekDaylight  TimeWeekDaylight_EPI  \
LineKeyOrigin                                                              
00000837B868F926F67BD440A0845C1D                47                     0   
00001ADC8C5705AB4FD8197B5024C1AA               744                     0   
00001C544BEBBB9109CE3552B4CDC490                 0                     0   
00002AFB398A29AE5AA1729B78DE22EB               104                     0   
000049FDC90665F2A02CCD1F5FABD4D8                 0                     0   

                                  TimeWeekNight  TimeWeekNight_EPI  \
LineKeyOrigin                                                        
00000837B868F926F67BD440A0845C1D             74                  0   
00001ADC8C5705AB4FD8197B5024C1AA            207                  0   
00001C544BEBBB9109CE3552B4CDC490             36                  0   
00002AFB398A29AE5AA1729B78DE22EB           5811                  0   
000049FDC90665F2A02CCD1F5FABD4D8              0                  0   

                                  TimeWeekend  TimeWeekend_EPI  
LineKeyOrigin                                                   
00000837B868F926F67BD440A0845C1D          109                0  
00001ADC8C5705AB4FD8197B5024C1AA          218                0  
00001C544BEBBB9109CE3552B4CDC490            0                0  
00002AFB398A29AE5AA1729B78DE22EB         1468                0  
000049FDC90665F2A02CCD1F5FABD4D8           90                0

In [199]:
test.shape, test3.shape, test2.shape

((1927239, 14), (1215111, 14), (712128, 12))

In [ ]:
start_time = time.time()

new_cols =[col+"_EPI" for col in test.columns.values \
           if (col != 'EPIDEMIC' and not('Line' in col) and not('ANTENNA' in col) )]
#seteo todas las nuevas columnas en 0 y dropeo EPIDEMIC
only_non_epidemic_users = list(set(table[table['EPIDEMIC']==0].LineKeyOrigin.values) - set(table[table['EPIDEMIC']==1].LineKeyOrigin.values))

#test 2 se lleva los usuarios mixtos y test3 los "puros" non_epi
test2 = table[~table['LineKeyOrigin'].isin(only_non_epidemic_users)].copy()
test3 = table[table['LineKeyOrigin'].isin(only_non_epidemic_users)].copy()

for col in new_cols:
    test3[col]=0
    #test3.loc[test['LineKeyOrigin'].isin(only_non_epidemic_users),col]=0
test3.drop('EPIDEMIC',inplace=True,axis=1)
#paso el linekeyorigin a indice y dropeo esa columna
test3 = test3.groupby('LineKeyOrigin').sum()
# uso el agregador en el resto de los users
print("finished setting pure epidemic users, current time is {t} \n".\
              format(t=(time.time()-start_time))) 
test2 =test2.groupby('LineKeyOrigin').apply(get_epidemic_agregator)
#seteo bien el indice a LineKeyOrigin y dropeo el level 1 del multi
test2.reset_index(level=1,drop=True,inplace=True)
#finalmente concateno las dos tablas
test = pd.concat([test3,test2])

elapsed_time = time.time()-start_time
print("total algo time for {n} users is {t}".format(n=(len(test2)+len(test3)), t= elapsed_time))

### Sanity checking im not losing users in the algorithm

In [207]:
table.shape, test.shape, test2.shape,test3.shape

((5015519, 11), (1927239, 20), (712128, 12), (1215111, 20))

In [208]:
1927239== 712128 +1215111

True

In [231]:
print(table.shape)
print(table[~table['LineKeyOrigin'].isin(only_non_epidemic_users)].shape)
print(table[table['LineKeyOrigin'].isin(only_non_epidemic_users)].shape)

(5015519, 11)
(1862025, 11)
(3153494, 11)


1927239

In [228]:
table['LineKeyOrigin'].isin(only_non_epidemic_users).isnull()

Series([], Name: LineKeyOrigin, dtype: object)

In [222]:
null_rows= table[(table[relevant_cols].T ==0).all()].index
len(null_rows)

0

In [ ]:
non_epi_hashes =np.unique(table[~table['LineKeyOrigin'].isin(only_non_epidemic_users)].LineKeyOrigin.values)
epi_hashes = np.unique(table[table['LineKeyOrigin'].isin(only_non_epidemic_users)].LineKeyOrigin.values)
len(epi_hashes) + len(non_epi_hashes)

In [233]:
all_hashes =np.unique(table.LineKeyOrigin.values)
print len(all_hashes)
print len(epi_hashes) + len(non_epi_hashes)

1927239
1927239


In [219]:
hashes = np.unique(user_epidemic.index.values)
len(hashes)

3667451

## Results
Levantamos en sum_links 1927239 usuarios de la Telco que tienen al menos 1 interaccion con otro users de la misma TelCo
Sin embargo en la tabla de simple_format ya procesada, tengo 3667451 users de la Telco.
Lo que creo que esta sucediendo es que hay 1.5M de users que NO hablan con otro usuario de la telco.... Raro.

Habria que testear en simple_format de levantar de vuelta los useres pertinentes asi nos

### Sanity checking im not losing users
hicimos una comparacion de la tabl antes y despues de la ultima linea de codigo y da True en

len(np.unique(table.LineKeyOrigin.values)) == 640647

len(np.unique(test.index.levels[0].values)) == 640647

In [150]:
a.index

Int64Index([7], dtype='int64')

In [153]:
test2.head(7:14).apply(lambda x: x,axis=1)

SyntaxError: invalid syntax (<ipython-input-153-693557ea2e74>, line 1)

In [ ]:
## script para convertir las columnas de otros types --> ints

passes=5

#ver el tiempo que tarda
now = time.time()

#el chunk basicamente va leyendo el file de a 'chunksize' cantidad de filas
#subgroup = pd.DataFrame()

for group in range(passes):
    subgroup = pd.DataFrame()
    table = pd.read_csv(
            input_file,
            engine = 'c',
            chunksize = 5*10**7,
#            iterator =True,
            sep = ' ',
            header = None,
            index_col=None,
            names = ['Target','Destination','Direction','TimeStamp','Duration','AntennaID'],
            usecols = ['Target', 'AntennaID','TimeStamp'],
            dtype = {'Target':np.uint32,'Destination':np.uint32,'TimeStamp':np.uint32,
                                      'Duration':np.uint16,'AntennaID':np.uint16,'Direction':np.object_}
            )
    
    #cuando entramos a este loop, table tiene tantos 'chunks' como el valor entero de la cantidad de lineas en el file
    #dividido el tamanyo del chunksize
    for chunk in table:
            #a cada chunk filtro por todos los Targets que tengan resto group modulo passes. y trabajo sobre la tabla
            #subgroup nada mas que ahora tiene pocos usuarios
            chunk['Target_int'] = chunk[].map(users_dict)
            subgroup = subgroup.append(chunk[chunk['Target'] % passes == group])
    #
    #entonces la idea es que yo ahora solo voy a trabajar, dentro de la tabla
    
    #paso de segundos a horas
    #notar que timestamp arranca en 0 segundos para domingo 01/01/2012 00:00am 
    #con lo cual domingo es el dia 0, lunes el 1, asi..
    
    subgroup['Hour'] =  (subgroup['TimeStamp'].values*1.0/3600)%24
    subgroup['Day']  =  (subgroup['TimeStamp'].values*1.0/(3600*24))%7
    
    #filtro usuarios con pocos o demasiados llamados en general menos de 5 mensuales y mas de 400  
    insignificant_users = subgroup['Target'].value_counts()[(subgroup['Target'].value_counts() < 5) \
                        | (subgroup['Target'].value_counts() > 400)].index.values.tolist()
    subgroup= subgroup.loc[~subgroup['Target'].isin(insignificant_users)]
    
    
    #filtro segun nightfilter y week_end
    if night_filter == True:
        subgroup = subgroup.loc[(subroup['Hour']<8) | (subgroup['Hour']>18)]
    if week_end == True:
        subgroup = subgroup.loc[(subgroup['Day']==0) | (subgroup['Day']==6)]

    
    #   
        
    grouped = subgroup.groupby(['Target', 'AntennaID'])['AntennaID'].agg({'count': np.size})
    grouped.reset_index(inplace=True,drop=False)
    
    del subgroup
    #reordeno dentro de c/ target por el count del antenna, esto me sirve para despues ordenar las antenas por uso
    grouped.sort_values(by=['Target','count'],ascending=False,inplace=True)
    
    ##enriquezco la muestra con datos epidemicos
    #primero agrego a cada antenna del df el dato de si es epidemica
    #despues agrupo por el target y me fijo solo la columna epidemica en c/grupo
    #finalmente sumo en c/ grupo y tomo la parte superior 
#entera de esa division con el largo del grupo. Si uso al menos una antena epidemica entonces esta expuesto(==1) Si no,
# da 0 pues no estuvo expuesto.        
    ##enriquezco la muestra con datos epidemicos

    exposed_info =grouped.join(antennas['EPIDEMIC'], on='AntennaID').\
    groupby('Target')['EPIDEMIC'].\
        agg({'EXPOSED' : lambda x: int( np.ceil(np.sum(x)*1.0/np.size(x)) )}) 
    
    #actualizo la tabla
    grouped = grouped.join(exposed_info['EXPOSED'],on="Target")
    del exposed_info
    
    
    #creo la tabla filtrada solo por users, que es la que voy a terminar guardando (hay tantos rows como users)
    output_table = grouped.drop_duplicates(subset = 'Target', keep='first')
    #re indexo
    output_table.index = output_table['Target'].values
    
    #agrupo ahora la tabla por target para hacer todos los calculos en los grupos
    
    grouped = grouped.groupby('Target')  
    
    #aca voy a ir agregando las top 10 antennas utilizadas por el user, Si no llego a 10 antennas, relleno con NaNs
    for i in range(0,10):
        #me quedo con la i-esima fila de c/grupo (si no hay fila, no toma en cuenta ese Target)
        
        buffer_table = grouped.nth(i)[['AntennaID','count']]
        #renombre a iesima AntennaId e iesimo count
        buffer_table.columns=['AntennaID_%i'%i,'count_%i'%i]
        #agrego esta info como nuevas columnas, dejando lo demas como NaNs
        output_table = pd.concat([output_table, buffer_table], axis=1, join_axes=[output_table.index])
    
    del grouped
    #los primeros AntennaIDs ya no me sirven, idem con el primer
    output_table.drop('AntennaID', axis=1, inplace=True)
    output_table.drop('count', axis=1, inplace=True)
    
    #para los datos faltantes dentro del Top10, relleno con -1s, que vendrian a ser los NaNs
    output_table.fillna(-1,inplace=True)
    
    #ojo aca que en el caso que entren hashes entonces no los va a poder convertir
    output_table = output_table.astype(int,copy=False)
    
    #agrego info de EPIDEMIC, asumiendo a alguien como epidemico si al antenna donde vive (la AntennaID_0) 
    #esta catalogada como EPIDEMIC
    test_table =test_table.join(antennas['EPIDEMIC'], on='AntennaID_0')
    
    
    #print(output_table.columns)
    #print("la tabla es:\n") 
    #print(output_table.head(5))
    #print("\n")
    
    #print(output_table.dytpes)
    #pd.to_numeric(output_table['EXPOSED'])
    
    #for i in range(3,22):
    #    pd.to_numeric(output_table[output_table.columns[i]]) 
    #print(output_table.dytpes)
    
    
    
    
    #aca termino guardando el output final pero solo para esos usuarios % pass ==group
    output_table.to_csv(output_file, index = False, 
                   header = False, mode='a',compression='gzip')
then = now
seconds = time.time() - then
print("total running time of script is %d " % seconds)

In [ ]:
#para observar la distribucion de veces que un usuario llama en un mes.
tabla['Target'].value_counts()[(tabla['Target'].value_counts()>5) & (tabla['Target'].value_counts() < 500)].hist(bins=100)

In [ ]:
sample['Target'].value_counts()[(sample['Target'].value_counts()>5) & (sample['Target'].value_counts() < 500)].hist(bins=35)

In [ ]:
#con esto extraigo el mapping user --> antennas used that month
user_antenna = tabla.groupby('Target')['AntennaID'].apply(lambda x: x.tolist()).to_dict();
user_antenna2 = tabla.groupby('Target')['AntennaID'].apply(lambda x: x.tolist()).to_dict();

user_antenna_map_concat = defaultdict(list)

#Veo como concatenar dos "maps" distintos y dejando solo las antenas unicas..
for d in (user_antenna_map, user_antenna_map2): # podemos meter cuantos dicts queramos aca..
    for key, value in d.iteritems():
        #list(set()) elimina repetidos
        user_antenna_map_concat[key] = list(set(user_antenna_map_concat[key] + value))


In [ ]:
import subprocess
subprocess.call(" echo \"Hello World \" ",shell=True)

In [ ]:
os.listdir(os.getcwd())    

In [ ]:
os.chdir(rootdir)
os.walk('.').next()[1]

In [ ]:
os.chdir(rootdir)
print "testdata" in os.listdir(os.getcwd()) 
if "testdata" in os.listdir(os.getcwd()):
    os.system("mkdir testdata")

In [ ]:
os.getcwd()

In [ ]:
os.listdir(os.getcwd())

In [ ]:
!ls